### PR TITLE
[_362] escape special characters in PAM passwords

### PR DESCRIPTION
**Pull request by d-w-moore**
Adaptation of solution per @lwesterhof  for #362

### CODE REVIEW

**Comment by trel** [comment]
Seems good - @lwesterhof any concerns / suggestions?

**Comment by lwesterhof** [comment]
The `PAM_PW_ESC_PATTERN` is missing the semicolon I think. This PR works instead of our workaround, but not for passwords with a semicolon.

**Comment by d-w-moore** [comment]
Yes and I'll  add a test, if we think we need it.

**Comment by trel** [comment]
Yes, please add a test to exercise the new fancy thing.

**Comment by trel** [comment]
please squash for review when ready

**Comment by d-w-moore** [comment]
yep will be first priority other than 4.3.0 activities

**Comment by trel** [comment]
seems good - please squash to three.

**Comment by d-w-moore** [comment]
squashed . ready for review, but test is still forthcoming

File: irods/connection.py [hunk 1]
```diff
@@ -290,18 +290,22 @@ def disconnect(self):
...
+                try:
+                    # SSL shutdown handshake
+                    self.socket = self.socket.unwrap()
+                except AttributeError:
+                    pass
+                self.socket.shutdown(socket.SHUT_RDWR)
+                self.socket.close()
+        finally:
+            self._disconnected = True  # Issue 368 - because of undefined destruction order during interpreter shutdown,
+            self.socket = None         # as well as the fact that unhandled exceptions are ignored in _del__, we'd at least
```
Comment (alanking) [inline irods/connection.py:306]: _del__ -> __del__

File: irods/test/login_auth_test.py [hunk 2]
```diff
@@ -118,7 +118,8 @@ class TestLogins(unittest.TestCase):
          Per https://slides.com/irods/ugm2018-ssl-and-pam-configuration#/3/7
 
       4. Create rodsuser alissa and corresponding unix user with the appropriate
-         passwords as below.
+         passwords as below.  For test_0[1-4] and test[1-4] to pass, user alissa must
+         have the native iRODS password set, ie.: 'iadmin moduser alissa password apass'.
```
Comment (alanking) [inline irods/test/login_auth_test.py:122]: For 4.3.0+, the pam authentication scheme should be updated to pam_password (see line 131 and 162).

I'm also a little surprised that it's required to have a native authentication password set, but I don't see a need to look into this right now.

**Review by alanking — COMMENTED** [review]
Looks good for the most part. If this is supposed to compatible with the pam_password plugin, we will need to (possibly selectively) replace references to the "pam" authentication scheme to "pam_password".

**Review by alanking — COMMENTED** [review]
Ah, y'know what, I just remembered that this is PRC and it is not using the C++ plugin-driven flow. Please disregard my previous comments.

Awaiting test before approval

File: irods/test/login_auth_test.py (hunk 2 above)
Comment (alanking) [reply irods/test/login_auth_test.py:122]: Please disregard.

**Review by korydraughn — COMMENTED** [review]
Looks fine to me.

Does the member variable `pam_password` have anything to do with the new pam_password authentication flow/lib?

**Comment by d-w-moore** [comment]
@korydraughn  The `pam_password` member is merely the escaped version of the `password `member.  It probably needn't exist; the same could be done with a variable local to the _login_pam( ) method.

**Comment by korydraughn** [comment]
If they aren't needed and are trivial to remove, then remove them. Else, capture it in an issue so we can handle it later.

**Comment by lwesterhof** [comment]
> one more vote of confidence from @lwesterhof and I think we're good to squash, #, and merge.

LGTM, tested it again and the PR works as replacement of our workaround.

File: irods/connection.py (hunk 1 above)
Comment (d-w-moore) [reply irods/connection.py]: done

**Comment by d-w-moore** [comment]
@trel @korydraughn - https://github.com/irods/python-irodsclient/issues/370 created, so we can admit this PR asap for release v1.1.4
> If they aren't needed and are trivial to remove, then remove them. Else, capture it in an issue so we can handle it later.

**Comment by trel** [comment]
please rebase so they apply cleanly.

**Comment by d-w-moore** [comment]
done. passes tests. final eyes-on time ...

**Comment by trel** [comment]
i like it.   please # em.

**Comment by d-w-moore** [comment]
#'d

**Pull request merged**